cpp/src/arrow/flight/client.cc

```diff
@@ -148,7 +142,28 @@ class GrpcClientInterceptorAdapter : public grpc::experimental::Interceptor {
   }
 
  private:
+  void ReceivedHeaders(std::multimap<grpc::string_ref, grpc::string_ref>* metadata) {
+    if (received_headers_) {
```

Comment:
Yes - I did it that way as we don't normally expose trailers in other contexts (e.g. there's no way to send one), so I decided it would be best to not mix the two except when we're forced to by grpc-java.

(Or rather, we do expose them with custom error metadata, but that's in a limited context.)